File: python system/inventory_system/repositories/transaction_repository.py
from typing import List, Optional
from datetime import datetime
from models.transaction import Transaction, TransactionType
from repositories.database import InMemoryDatabase


class TransactionRepository:
    def __init__(self, db: InMemoryDatabase):
        self._db = db

    def get_all(self) -> List[Transaction]:
        return self._db.transactions

    def get_by_item_id(self, item_id: str) -> List[Transaction]:
        return self._db.get_transactions_for_item(item_id)

    def get_by_type(self, type: TransactionType) -> List[Transaction]:
        return [t for t in self._db.transactions if t.type == type]

    def get_by_date_range(self, from_date: datetime, to_date: datetime) -> List[Transaction]:
        return [
            t for t in self._db.transactions
            if from_date < t.timestamp < to_date
        ]

    def save(self, transaction: Transaction) -> None:
        self._db.insert_transaction(transaction)

File: dart/transaction_system/dart_application_1/bin/transaction_repository.dart
// lib/repositories/transaction_repository.dart

import 'models/transaction.dart';
import 'database.dart';

class TransactionRepository {
  final InMemoryDatabase _db;

  TransactionRepository(this._db);

  List<Transaction> getAll() => _db.transactions;

  List<Transaction> getByItemId(String itemId) =>
      _db.getTransactionsForItem(itemId);

  List<Transaction> getByType(TransactionType type) =>
      _db.transactions.where((t) => t.type == type).toList();

  List<Transaction> getByDateRange(DateTime from, DateTime to) =>
      _db.transactions
          .where((t) =>
              t.timestamp.isAfter(from) && t.timestamp.isBefore(to))
          .toList();

  void save(Transaction transaction) => _db.insertTransaction(transaction);
}
